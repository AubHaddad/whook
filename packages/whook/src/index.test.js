import { constant, initializer } from 'knifecycle';
import {
  runServer,
  prepareServer,
  prepareEnvironment as basePrepareEnvironment,
} from './index';

describe('runServer', () => {
  it('should work', async () => {
    const PORT = 8888;
    const HOST = 'localhost';
    const BASE_PATH = '/v1';
    const API = {
      openapi: '3.0.2',
      info: {
        version: '1.0.0',
        title: 'Sample OpenAPI',
        description: 'A sample OpenAPI file for testing purpose.',
      },
      paths: {
        '/ping': {
          head: {
            operationId: 'getPing',
            summary: "Checks API's availability.",
            responses: {
              '200': {
                description: 'Pong',
              },
            },
          },
        },
      },
    };
    const logger = {
      info: jest.fn(),
      error: jest.fn(),
    };
    const debug = jest.fn();
    const $autoload = jest.fn(async () => ({}.undef));

    async function prepareEnvironment() {
      const $ = await basePrepareEnvironment();

      $.register(
        initializer(
          {
            name: '$autoload',
            type: 'service',
            options: { singleton: true },
          },
          async () => $autoload,
        ),
      );
      $.register(constant('BASE_PATH', BASE_PATH));
      $.register(constant('API', API));
      $.register(constant('ENV', {}));
      $.register(constant('NODE_ENV', 'test'));
      $.register(constant('PORT', PORT));
      $.register(constant('HOST', HOST));
      $.register(constant('WRAPPERS', []));
      $.register(constant('DEBUG_NODE_ENVS', []));
      $.register(constant('NODE_ENVS', ['test']));
      $.register(
        constant('HANDLERS', {
          getPing: jest.fn(() => ({ status: 200 })),
        }),
      );
      $.register(constant('logger', logger));
      $.register(constant('debug', debug));

      return $;
    }
    process.env.ISOLATED_ENV = 1;

    const { $destroy } = await runServer(prepareEnvironment, prepareServer, [
      '$destroy',
      'httpServer',
      'process',
    ]);

    await $destroy();

    expect({
      debugCalls: debug.mock.calls,
      logInfoCalls: logger.info.mock.calls,
      logErrorCalls: logger.error.mock.calls,
      autoloaderCalls: $autoload.mock.calls,
    }).toMatchSnapshot();
  });
});
