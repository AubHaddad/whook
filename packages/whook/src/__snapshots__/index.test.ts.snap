// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runServer should work 1`] = `
Object {
  "debugCalls": Array [
    Array [
      "⌛ - Delay service initialized.",
    ],
    Array [
      "⏰ - Time service initialized.",
    ],
    Array [
      "⏳ - Cancelling pending timeouts:",
      0,
    ],
    Array [
      "✅ - Closing HTTP server.",
    ],
    Array [
      "✔️ - HTTP server closed!",
    ],
    Array [
      "❤️ - Initializing the APM service.",
    ],
    Array [
      "👣 - Logging service initialized.",
    ],
    Array [
      "💱 - HTTP Transaction initialized.",
    ],
    Array [
      "📇 - Process service initialized.",
    ],
    Array [
      "🕶️ - Initializing the obfuscator service.",
    ],
    Array [
      "🚦 - HTTP Router initialized.",
    ],
  ],
  "logErrorCalls": Array [
    Array [
      "🔂 - Running in \\"test\\" environment.",
    ],
    Array [
      "🎙️ - HTTP Server listening at \\"http://localhost:8888\\".",
    ],
    Array [
      "On air 🚀🌕",
    ],
  ],
  "logInfoCalls": Array [],
}
`;
